tests/federation_upload_keys_test.go

```diff
@@ -80,9 +83,20 @@ func TestFederationKeyUploadQuery(t *testing.T) {
 
 	// sytest: Can query remote device keys using POST
 	t.Run("Can query remote device keys using POST", func(t *testing.T) {
-		// Device list changes only come down incremental syncs, so we do an
-		// initial sync up front.
-		_, nextBatch := bob.MustSync(t, client.SyncReq{})
+		// We expect the key upload to come down /sync. We need to do this so
+		// that can tell the next device update actually triggers the
+		// notification to go down /sync.
+		nextBatch := bob.MustSyncUntil(t, client.SyncReq{Since: nextBatchBeforeKeyUpload}, func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+			devicesChanged := topLevelSyncJSON.Get("device_lists.changed")
+			if devicesChanged.Exists() {
+				for _, userID := range devicesChanged.Array() {
+					if userID.Str == alice.UserID {
+						return nil
+					}
+				}
+			}
```

Comment:
there might be some clever way to use JSONCheckOffAllowUnwanted here but I usually find that more confusing tbh